### ISSUE TITLE
False positive no-access-state-in-setstate when using this.state in a function called in the setstate callback

### ISSUE THREAD

**Issue by Aequitas737** [yannickcr/eslint-plugin-react#1931]
Related to #1610, which doesn't completed catch all the cases.

With version 7.10.0 when the callback calls a function which references this.state.

```
myFunc = () => {
   this.setState({
            ...prop,
        }, () => this.doStuff());
    }

    doStuff= () => {
        console.log(this.state.asdf);
    }
```

Can work around this by just passing this.state into doStuff: `this.doStuff(this.state)`

**Comment by ljharb** [comment]
You definitely don't ever want to pass around the props or state object; we need to get this fixed.

**Comment by eric-wood** [comment]
Also seeing this, and have found I can work around it by using destructuring:

```js
this.setState({ ...stuff }, () => {
  const { asdf } = this.state;
  console.log(asdf);
});
```

**Comment by jaybekster** [comment]
Has it been fixed?

**Comment by ljharb** [comment]
@jaybekster nope, that's why the issue is still open

**Comment by Jexah** [comment]
+1 for interest in fixing.

**Comment by jzabala** [comment]
Working properly with versions:

```
"babel-eslint": "^10.1.0"
"eslint": "^7.4.0"
"eslint-plugin-react": "^7.20.3"
```

![Screen Shot 2020-07-11 at 11 12 09 AM](https://user-images.githubusercontent.com/1315054/87227123-c35db400-c366-11ea-83d3-36c4b8848e71.png)
